extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,298 @@
+
+      // Note: The parsing for variables breaks when `|` is used as
+      // something other than the filter separator. A full-fledged parser would
+      // be needed to fix this.
+      const tokens = key.split('|');
+      const initialValue = tokens.shift().trim();
+      if (!initialValue) {
+        return Promise.resolve('');
+      }
+
+      const {name, argList} = this.getNameArgs_(initialValue);
+      const raw = options.vars[name] || '';
+
+      let p;
+      if (typeof raw == 'string') {
+        // Expand string values further.
+        p = this.expandTemplate(raw,
+            new ExpansionOptions(options.vars, options.iterations - 1,
+                options.noEncode));
+      } else {
+        // Values can also be arrays and objects. Don't expand them.
+        p = Promise.resolve(raw);
+      }
+
+      p = p.then(expandedValue =>
+          // First apply filters
+          this.applyFilters_(expandedValue, tokens))
+        .then(finalRawValue => {
+          // Then encode the value
+          const val = options.noEncode
+              ? finalRawValue
+              : this.encodeVars(finalRawValue, name);
+          return val ? val + argList : val;
+        })
+        .then(encodedValue => {
+          // Replace it in the string
+          replacement = replacement.replace(match, encodedValue);
+        });
+
+      // Queue current replacement promise after the last replacement.
+      replacementPromises.push(p);
+
+      // Since the replacement will happen later, return the original template.
+      return match;
+
+    });
+
+    // Once all the promises are complete, return the expanded value.
+    return Promise.all(replacementPromises).then(() => replacement);
+  }
+
+  /**
+   * Returns an array containing two values: name and args parsed from the key.
+   *
+   * @param {string} key The key to be parsed.
+   * @return {!Object<string>}
+   * @private
+   */
+  getNameArgs_(key) {
+    if (!key) {
+      return {name: '', argList: ''};
+    }
+    const match = key.match(VARIABLE_ARGS_REGEXP);
+    user().assert(match, 'Variable with invalid format found: ' + key);
+    return {name: match[1] || match[0], argList: match[2] || ''};
+  }
+
+  /**
+   * @param {string|!Array<string>} raw The values to URI encode.
+   * @param {string} unusedName Name of the variable.
+   * @return {string} The encoded value.
+   */
+  encodeVars(raw, unusedName) {
+    if (!raw) {
+      return '';
+    }
+
+    if (isArray(raw)) {
+      return raw.map(encodeURIComponent).join(',');
+    }
+    // Separate out names and arguments from the value and encode the value.
+    const {name, argList} = this.getNameArgs_(String(raw));
+    return encodeURIComponent(name) + argList;
+  }
+
+  /**
+   * @param {string} value
+   * @return {!Promise<!string>}
+   */
+  hashFilter_(value) {
+    return cryptoFor(this.win_).then(crypto =>
+      crypto.sha384Base64(user().assertString(value)));
```

Comment:
This filter is declared with `allowNulls=false`, thus `value` can no longer be `null` here, right? So we don't need `assertString`?